server/queue/queue-jms/src/test/java/org/apache/james/queue/jms/JMSMailQueueTest.java

```diff
@@ -49,7 +49,8 @@ public void setUp(BrokerService broker) throws Exception {
         RawMailQueueItemDecoratorFactory mailQueueItemDecoratorFactory = new RawMailQueueItemDecoratorFactory();
         NoopMetricFactory metricFactory = new NoopMetricFactory();
         String queueName = BrokerExtension.generateRandomQueueName(broker);
-        mailQueue = new JMSMailQueue(connectionFactory, mailQueueItemDecoratorFactory, queueName, metricFactory);
+        ConsumerOptions consumerOptions = name -> name + "?consumer.prefetchSize=0";
```

Comment:
> From my point of view our configuration/options should not affect entire connection to the message broker. We should be able to configure queue consumer creation (ConsumerOptions#applyForDequeue).

I'm not arguing for changing the implementation much, just proposing to make the **ConsumerOptions** a dumb interface:

```
interface ConsumerOptions {
    ImmutableMap<String, String> parameters();
}
```

(Not that this interface is just holding a configuration and enforce no knowledge about the way the queue works.)

Then, We could get a JMSConsumerOptions, and an ActiveMQConsumerOptions with the options I can use in these various case.

Then, JMSMailQueue will add the right parameters for the queue name when appropriate (it has the knowledge of how to encode the parameters).

Thus:
 - We get strong typing for JMSConsumerOptions
 - The configuration is a dumb POJO, which is easier to reason about (and classic in James).

Would it be OK for you?

Reply:
Yes, in fact its a good idea, solving multiple problems/inconsistencies with a single shot.

Please, clarify how do you see the `JMSConsumerOptions` object creation? It would be through `new` or through builder class?

What do you think about both options combined together (public constructor + builder) ?

In case of builder, I suppose we should expose single `put(String, String)` and provide users context free consumer options builder.

Reply:
> Please, clarify how do you see the JMSConsumerOptions object creation? It would be through new or through builder class?

Through a builder, no? 

> What do you think about both options combined together (public constructor + builder) ?

Why using a constructor when I can use a builder? Isn't it better to enforce builder use?

> In case of builder, I suppose we should expose single put(String, String) and provide users context free consumer options builder.

The problem with that kind of API is that I got not idea about values I should pass. We can still expose the `put` you suggest, but we need a self-discoverable API for options we know.

WDYT?

Reply:
> The problem with that kind of API is that I got not idea about values I should pass. We can still expose the put you suggest, but we need a self-discoverable API for options we know.

Agree with that, but as far as I know JMS spec does not support things like that. Although we can provide pretty well documentation on that. In ActiveMQ case the builder API is pretty straightforward and easy do understand (from API design point of view).

> Why using a constructor when I can use a builder? Isn't it better to enforce builder use?

I thought about when user already has ready . Probably we should expose also the `putAll(Map<String, String>)` method in builder to not force user iterate over and add options one by one.